Source/WebCore/platform/gstreamer/GStreamerQuirkBroadcomBase.h

```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright (C) 2024 Igalia S.L
+ * Copyright (C) 2024 Metrological Group B.V.
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Library General Public
+ * License as published by the Free Software Foundation; either
+ * version 2 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Library General Public License for more details.
+ *
+ * You should have received a copy of the GNU Library General Public License
+ * aint with this library; see the file COPYING.LIB.  If not, write to
+ * the Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA 02110-1301, USA.
+ */
+
+#pragma once
+
+#if USE(GSTREAMER)
+
+#include "GStreamerCommon.h"
+#include "GStreamerQuirks.h"
+#include "MediaPlayerPrivateGStreamer.h"
+#include <wtf/Vector.h>
+
+namespace WebCore {
+
+class GStreamerQuirkBroadcomBase : public GStreamerQuirk {
+public:
+    GStreamerQuirkBroadcomBase();
+
+    bool needsBufferingPercentageCorrection() const { return true; }
+    bool queryBufferingPercentage(MediaPlayerPrivateGStreamer*, const char*& elementName, GRefPtr<GstQuery>&) const;
+    int correctBufferingPercentage(MediaPlayerPrivateGStreamer*, int originalBufferingPercentage, GstBufferingMode) const;
+    void resetBufferingPercentage(MediaPlayerPrivateGStreamer*, int bufferingPercentage) const;
+    void setupBufferingPercentageCorrection(MediaPlayerPrivateGStreamer*, GstState currentState, GstState newState, GstElement*) const;
+
+private:
+    class MovingAverage {
+    public:
+        MovingAverage(size_t length)
+            : m_values(length)
+        {
+        }
+
+        void reset(int value)
+        {
+            for (size_t i = 0; i < m_values.size(); i++)
+                m_values[i] = value;
+        }
+
+        int accumulate(int value)
+        {
+            intmax_t sum = 0;
+            for (size_t i = 1; i < m_values.size(); i++) {
+                m_values[i - 1] = m_values[i];
+                sum += m_values[i - 1];
+            }
+            m_values[m_values.size() - 1] = value;
+            sum += value;
+            intmax_t result = sum / m_values.size();
+            ASSERT(result < std::numeric_limits<int>::max());
```

Comment:
I think this overflow check is wrong. If I am not mistaken, if you add two numbers and it overflows, what you are going to get is a negative number so always smaller than max.

Anyway, from what I understand, you are adding natural numbers so first, you can use uintmax_t. Then to check the overflow, you just check if the sum of the two numbers is smaller than one of them, then you know you overflew.

Reply:
No, no. Have into account that I'm using intmax_t for the operations (which I hope is much larger than the regular int type and won't ever overflow), but then I want to fit that result in the regular int type. I perform the check in "the intmax_t world".

Now let's consider some actual numbers: The default MovingAverage size used in this patch is 10. That means a maximum of 10 * 100 = 1000 accumulated value. To make this computation overflow in a 64 bit signed int, we would need a MovingAverage with 9223372036854775807 / 100 = 92233720368547758 items. No computer can hold that array in memory in the foreseable future (famous last words, when computers had 640KB of RAM).

Maybe I can add a constructor-time assert to ensure that the caller isn't ever trying to construct a MovingAverage with more than, let's say, 50 elements, and also ensure that values can't be greater than 100. If I do that, I could even avoid the need of intmax_t.